test/malli/roundtrip_test.clj

```diff
@@ -17,49 +18,361 @@
     (is (nil? (rt/explain-roundtrip [:nil])))
     (is (nil? (rt/explain-roundtrip [:any])))
-    (is (nil? (rt/explain-roundtrip [:number])))
-    (is (nil? (rt/explain-roundtrip [:enum "a" "b" "c"]))))
+    (is (nil? (rt/explain-roundtrip number?)))
+    (is (nil? (rt/explain-roundtrip int?)))
+    (is (nil? (rt/explain-roundtrip [:enum "a" "b" "c"])))))
 
 (deftest roundtrippable-or-nonoverlapping
   (testing ":or with non-overlapping branches is roundtrippable"
     (is (nil? (rt/explain-roundtrip [:or [:int] [:string]])))
     (is (nil? (rt/explain-roundtrip [:or [:keyword] [:string]])))))
 
 (deftest roundtrippable-or-overlapping
-  (testing ":or with overlapping branches is not roundtrippable"
-    (let [result (rt/explain-roundtrip [:or [:int] [:number]])]
-      (is (vector? result))
-      (is (seq result))
-      (is (some #(re-find #"overlap" (:problem %)) result)))
-    (let [result (rt/explain-roundtrip [:or [:number] [:int]])]
-      (is (vector? result))
-      (is (seq result))
-      (is (some #(re-find #"overlap" (:problem %)) result)))))
+  (testing ":or with overlapping simple-parser branches IS roundtrippable"
+    ;; When both branches have simple (non-transforming) parsers,
+    ;; there's no ambiguity even if types overlap
+    (is (nil? (rt/explain-roundtrip [:or [:int] number?])))
+    (is (nil? (rt/explain-roundtrip [:or number? [:int]]))))
+  (testing ":or with overlapping non-simple branches is NOT roundtrippable"
+    ;; When branches have transforming parsers and overlap, it's not roundtrippable
+    (let [schema [:or [:orn [:i :int]] [:orn [:j :int]]]
+          result (rt/explain-roundtrip schema)]
+      ;; Verify the exact error structure
+      (is (= [{:schema schema
+               :branch-a [:orn [:i :int]]
+               :branch-b [:orn [:j :int]]
+               :path [0 1]
+               :problem ":or branches at positions 0 and 1 overlap in their parsed domain, so this schema is not roundtrippable. If you need roundtripping, use :orn instead of :or."}]
+             result))
+      ;; Note: This schema actually roundtrips if the first branch is chosen,
+      ;; but we flag it as potentially non-roundtrippable for soundness
+      )))
 
 (deftest roundtrippable-or-fn-orn
-  (testing ":or with :fn and :orn allows roundtripping"
-    (is (nil? (rt/explain-roundtrip
-                [:or [:fn record?] [:orn [:i :int]]])))))
+  (testing ":or with :fn record? and :orn does NOT allow roundtripping (order matters)"
+    ;; :orn produces Tag records, which match record? predicate
+    ;; When :fn record? comes first, it captures the Tag during unparse
+    (let [schema [:or [:fn record?] [:orn [:i :int]]]
+          v1 123
+          result (rt/explain-roundtrip schema)]
+      ;; Verify the error (note: fn shows as object in form)
+      (is (= 1 (count result)))
+      (is (= [0 1] (:path (first result))))
+      (is (= [:orn [:i :int]] (:branch-b (first result))))
```

Comment:
Please collapse this into one `(is (=))` test. I want to see the entire value. Try `[:fn record?]` as the expected `:branch-a` instead of `[:fn 'record?]`. `:fn` schemas return their children verbatim in `m/form`.